walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,114 @@
 
 package com.google.sps;
 
+import com.google.sps.Event;
+import com.google.sps.MeetingRequest;
+import com.google.sps.TimeRange;
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.lang.String;
+import java.io.*;
 
 public final class FindMeetingQuery {
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    
+    // Request is too long or too short(more than 1 day/negative duration)
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration() || request.getDuration() <= 0)
+    {
+      return Arrays.asList();
+    }
+
+    // Return full day for trivial empty cases (no event given, no attendees given)
+    if (events.isEmpty() || request.getAttendees().isEmpty())
+    {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    events = sortEventsByStartTime(events);
+
+    int requiredDuration = (int)request.getDuration(); // Can cast as such because duration does not exceed 2^32.
+    Collection<String> requiredAttendees = request.getAttendees();
+
+    ArrayList<TimeRange> candidateRanges = new ArrayList<TimeRange>();
+
+    Iterator eventsIterator = events.iterator();
+    Event currentEvent = (Event)eventsIterator.next();
+    Event previousEvent = currentEvent;
+    int currentTime = TimeRange.START_OF_DAY;
+    int startTime = TimeRange.START_OF_DAY;
+    
+    while (true) {
+
+      boolean attendeesNotIncludedInCurrentEvent = Collections.disjoint(currentEvent.getAttendees(), request.getAttendees());
+      
+      if (attendeesNotIncludedInCurrentEvent == false) 
+      { 
+        currentTime = currentEvent.getWhen().start(); // Available time is only up to when event starts, since there are clashing attendees.
+
+        TimeRange potentialSlot = TimeRange.fromStartEnd(startTime, currentTime, false);
+        if (potentialSlot.duration() >= requiredDuration)
+        {
+          candidateRanges.add(potentialSlot);
+        }
+
+       startTime = setStartTimeBasedOnPreviousEvent(previousEvent, currentEvent, request);
+      }
+
+      // Check for next event
+      if (eventsIterator.hasNext())
+      {
+        previousEvent = currentEvent;
+        currentEvent = (Event)eventsIterator.next();
+        continue;
+      }
+      else // Terminate
+      {
```

Comment:
You can remove the `else` here since the `if` statement above will be terminated by the `continue`.

Reply:
done!